File: lunar-backend/lunar-server/src/main/java/com/nevermore/lunar/server/vo/CommonResult.java
package com.nevermore.lunar.server.vo;

import static com.nevermore.lunar.server.vo.ResultCode.BAD_REQUEST;
import static com.nevermore.lunar.server.vo.ResultCode.INTERNAL_SERVER_ERROR;
import static com.nevermore.lunar.server.vo.ResultCode.SUCCESS;

/**
 * @author nevermore
 */
public record CommonResult<T>(
        int code,
        T result,
        String message
) {

    public static CommonResult<?> successResult() {
        return from(SUCCESS);
    }

    public static CommonResult<?> badRequestResult() {
        return from(BAD_REQUEST);
    }

    public static CommonResult<?> internalServerErrorResult() {
        return from(INTERNAL_SERVER_ERROR);
    }

    public static <T> CommonResult<T> successResult(T result) {
        return from(SUCCESS, result);
    }

    public static <T> CommonResult<T> from(ResultCode code) {
        return new CommonResult<>(code.getValue(), null, code.getMessage());
    }

    public static <T> CommonResult<T> from(ResultCode code, T result) {
        return new CommonResult<>(code.getValue(), result, code.getMessage());
    }
}

File: lunar-backend/lunar-server/src/main/java/com/nevermore/lunar/server/vo/CommonResult.kt
@file:Suppress("unused")

package com.nevermore.lunar.server.vo

import com.nevermore.lunar.server.vo.ResultCode.BAD_REQUEST
import com.nevermore.lunar.server.vo.ResultCode.INTERNAL_SERVER_ERROR
import com.nevermore.lunar.server.vo.ResultCode.SUCCESS

/**
 * @author nevermore
 */
sealed class LunarApiResult(
    val code: Int,
    val message: String
)

class CommonResult(code: Int, message: String) : LunarApiResult(code, message)

class DataResult<T>(code: Int, val data: T?, message: String) : LunarApiResult(code, message)

object ResultUtils {

    @JvmStatic
    @JvmOverloads
    fun successResult(message: String? = null) = from(SUCCESS, message)

    @JvmStatic
    @JvmOverloads
    fun badRequestResult(message: String? = null) = from(BAD_REQUEST, message)

    @JvmStatic
    @JvmOverloads
    fun internalServerErrorResult(message: String? = null) = from(INTERNAL_SERVER_ERROR, message)

    @JvmStatic
    fun <T> successResultWithData(data: T?): DataResult<T> = DataResult(SUCCESS.value, data, SUCCESS.message)

    @JvmStatic
    @JvmOverloads
    fun from(code: ResultCode, message: String? = null) = CommonResult(code.value, message ?: code.message)
}
